habitat-lab/habitat/tasks/nav/nav.py

```diff
@@ -1289,16 +1402,553 @@ def step(
+                        low=np.array([self._delta_ang_range_camera_pitch[0]]),
+                        high=np.array([self._delta_ang_range_camera_pitch[1]]),
+                        dtype=np.float32,
+                    ),
+                    "max_duration": spaces.Box(
+                        low=np.array([self._wait_duration_range[0]]),
+                        high=np.array([self._wait_duration_range[1]]),
+                        dtype=np.float32,
+                    ),
+                }
+            )
+
+    def step(
+        self,
+        task: EmbodiedTask,
+        xyt_waypoint: List[float],
+        delta_camera_pitch_angle: float,
+        max_duration: float,
+        *args,
+        **kwargs,
+    ):
+        # Preprocess waypoint input
+        assert len(xyt_waypoint) == 3, "Waypoint vector must be of length 3."
+        (
+            xyt_waypoint_processed,
+            delta_camera_pitch_angle_processed,
+            max_duration_processed,
+        ) = self._preprocess_action(
+            xyt_waypoint, delta_camera_pitch_angle, max_duration
+        )
+
+        # Execute waypoint
+        return self._step_rel_waypoint(
+            xyt_waypoint_processed,
+            delta_camera_pitch_angle_processed,
+            max_duration_processed,
+            *args,
+            **kwargs,
+        )
+
+    def _preprocess_action(
+        self, xyt_waypoint, delta_camera_pitch_angle, max_duration
+    ):
+        """Perform scaling and clamping of input"""
+        # Scale
+        if self._enable_scale_convert:
+            xyt_waypoint[0] = self._scale_inputs(
+                xyt_waypoint[0],
+                [-1, 1],
+                [
+                    self._waypoint_lin_range[0],
+                    self._waypoint_lin_range[1],
+                ],
+            )
+            xyt_waypoint[1] = self._scale_inputs(
+                xyt_waypoint[1],
+                [-1, 1],
+                [
+                    self._waypoint_lin_range[0],
+                    self._waypoint_lin_range[1],
+                ],
+            )
+            xyt_waypoint[2] = self._scale_inputs(
+                xyt_waypoint[2],
+                [-1, 1],
+                [
+                    self._waypoint_ang_range[0],
+                    self._waypoint_ang_range[1],
+                ],
+            )
+            delta_camera_pitch_angle = self._scale_inputs(
+                delta_camera_pitch_angle,
+                [-1, 1],
+                [
+                    self._delta_ang_range_camera_pitch[0],
+                    self._delta_ang_range_camera_pitch[1],
+                ],
+            )
+            max_duration = self._scale_inputs(
+                max_duration,
+                [0, 1],
+                [
+                    self._wait_duration_range[0],
+                    self._wait_duration_range[1],
+                ],
+            )
+
+        # Clamp
+        xyt_waypoint_clamped = np.array(
```

Comment:
If inputs are scaled, the clamping has different ranges.

Reply:
If `_enable_scale_convert is True`, then they would have been denormalized in the previous block of code.